infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+import org.corfudb.infrastructure.logreplication.transport.server.IServerChannelAdapter;
+import org.corfudb.runtime.LogReplication.LogReplicationSession;
+import org.corfudb.runtime.exceptions.NetworkException;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuInterruptedError;
+import org.corfudb.runtime.proto.service.CorfuMessage;
+import org.corfudb.util.CFUtils;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.IntervalRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getDefaultProtocolVersionMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getRequestMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getResponseMsg;
+
+/**
+ * A Base Source Router, which interfaces between the Log replication Source components (FSMs, AckReader, etc) and the
+ * custom transport adapter
+ *
+ * This class is inherited by both LogReplicationSourceClientRouter and LogReplicationSourceServerRouter.
+ */
+@Slf4j
+public class LogReplicationClientServerRouter implements IClientServerRouter {
+
+    public static String REMOTE_LEADER = "REMOTE_LEADER";
+
+    /**
+     * A map of remoteCluster to {@link CompletableFuture} for each session, which is completed when a connection,
+     * including a successful handshake, completes and messages can be sent
+     * to the remote node.
+     */
+    @Getter
+    private volatile Map<LogReplicationSession, CompletableFuture<Void>> sessionToLeaderConnectionFuture;
+
+    /**
+     * A map of session to the current requestID for the session.
+     */
+    @Getter
+    @SuppressWarnings("checkstyle:abbreviation")
+    private final Map<LogReplicationSession,AtomicLong> sessionToRequestIdCounter;
+
+    /**
+     * Sync call response timeout (milliseconds).
+     */
+    @Getter
+    @Setter
+    private long timeoutResponse;
+
+    /**
+     * The outstanding requests on this router.
+     */
+    @Getter
+    private final Map<LogReplicationSession, Map<Long, CompletableFuture>> sessionToOutstandingRequests;
+
+    /**
+     * Runtime FSM, to insert connectivity events
```

Comment:
same here, the comment and variable name reflect the fsm but it actually the runtime

Reply:
its the runtime FSM...
we have 2 FSMs on source right...1 is runtime FSM, and the other is the replication FSM.